src/psyclone/tests/psyir/nodes/array_mixin_test.py

```diff
@@ -297,86 +297,139 @@ def test_get_lbound_expression():
-        aref.get_lbound_expression(3)
+        aref._get_bound_expression(3, "lower")
     # Symbol is of DeferredType so the result should be an instance of the
     # LBOUND intrinsic.
     dtsym = DataSymbol("oops", DeferredType())
     dtref = ArrayReference.create(dtsym,
                                   [_ONE.copy(), _ONE.copy(), _ONE.copy()])
-    lbnd = dtref.get_lbound_expression(1)
+    lbnd = dtref._get_bound_expression(1, "lower")
     assert isinstance(lbnd, BinaryOperation)
     assert lbnd.operator == BinaryOperation.Operator.LBOUND
     assert lbnd.children[0].symbol is dtsym
     assert lbnd.children[1] == Literal("2", INTEGER_TYPE)
 
+    # Tests for ubound
+
+    # Symbol is of ArrayType.
+    ubound = DataSymbol("jmin", INTEGER_TYPE, is_constant=True,
+                        initial_value=Literal("10", INTEGER_TYPE))
+    ubnd_ref = Reference(ubound)
+    symbol = DataSymbol("my_symbol", ArrayType(INTEGER_TYPE,
+                                               [1, (1, 2), (1, ubnd_ref)]))
+    aref = ArrayReference.create(symbol,
+                                 [_ONE.copy(), _ONE.copy(), _ONE.copy()])
+    assert aref._get_bound_expression(0, "upper") == _ONE
+    assert aref._get_bound_expression(1, "upper") == _TWO
+    ub2 = aref._get_bound_expression(2, "upper")
+    assert isinstance(ub2, Reference)
+    # Returned lower bound should be a *copy* of the original.
+    assert ub2 is not ubnd_ref
+    assert ub2 == ubnd_ref
+
+    # Symbol is of DeferredType so the result should be an instance of the
+    # UBOUND intrinsic.
+    dtsym = DataSymbol("oops", DeferredType())
+    dtref = ArrayReference.create(dtsym,
+                                  [_ONE.copy(), _ONE.copy(), _ONE.copy()])
+    ubnd = dtref._get_bound_expression(1, "upper")
+    assert isinstance(ubnd, BinaryOperation)
+    assert ubnd.operator == BinaryOperation.Operator.UBOUND
+    assert ubnd.children[0].symbol is dtsym
+    assert ubnd.children[1] == Literal("2", INTEGER_TYPE)
+
 
 @pytest.mark.parametrize("extent", [ArrayType.Extent.DEFERRED,
                                     ArrayType.Extent.ATTRIBUTE])
-def test_get_lbound_expression_unknown_size(extent):
+def test_get_bound_expression_unknown_size(extent):
     '''
-    Test the get_lbound_expression method when we have the definition of the
+    Test the _get_bound_expression method when we have the definition of the
     array type but its dimensions are unknown.
 
     '''
     symbol = DataSymbol("my_symbol", ArrayType(INTEGER_TYPE,
                                                [extent, extent]))
     aref = ArrayReference.create(symbol, [_ONE.copy(), _ONE.copy()])
-    lbnd = aref.get_lbound_expression(1)
+    lbnd = aref._get_bound_expression(1, "lower")
     assert isinstance(lbnd, BinaryOperation)
     assert lbnd.operator == BinaryOperation.Operator.LBOUND
     assert lbnd.children[0].symbol is symbol
 
+    symbol = DataSymbol("my_symbol", ArrayType(INTEGER_TYPE,
```

Comment:
Can you just re-use the `symbol` and `aref` defined above?

Reply:
Yes